### PR TITLE
B2MD: Fix default memory reclamation strategy to match its documentation

### PR DESCRIPTION
**Pull request by fabianfett**
In cases in which only a small amount of bytes were consumed from a `ByteBuffer` during decoding a small error in our logic could lead to extensive reclaims of memory, degrading the overall performance and allocating way to much memory.

### Motivation:

- In cases in which only a small amount was decoded from the `ByteToMessageHandler`'s internal byteBuffer, we ran into reallocations based on our previous logic. If the byteBuffer was larger than 1kb and less than 50% were consumed we reallocated.
- This leads to performance problems
- But more importantly if within the decode function a byteBuffer slice is read, the internal `ByteToMessageHandler`'s byteBuffer's storage is now referenced at multiple locations a leading to fast and excessive memory growth.

### Modifications:

- A logic bug has been fixed, we reclaim memory now, if more than 50% of the byteBuffer has been read.
- The tests have been adjusted to reflect the fixed logic bug.

### Result:

- Faster decoding of small messages
- Less memory footprint when reading slices in small messages.

### CODE REVIEW

File: Sources/NIO/Codec.swift [hunk 1]
```diff
@@ -222,13 +222,13 @@ extension ByteToMessageDecoder {
...
-        // 1. If there is more than 2kB of memory to reclaim
+        // 1. If there is at least 2kB of memory to reclaim
         // 2. If the buffer is more than 50% reclaimable memory and is at least
         //    1kB in size.
-        if buffer.readerIndex > 2048 {
+        if buffer.readerIndex >= 2048 {
             return true
         }
-        return buffer.capacity > 1024 && (buffer.capacity - buffer.readerIndex) >= buffer.readerIndex
+        return buffer.capacity >= 1024 && (buffer.capacity - buffer.readerIndex) < buffer.readerIndex
```
Comment (weissi) [inline Sources/NIO/Codec.swift:231]: - the first change `>` --> `>=` essentially halves the threshold. We only allocate ByteBuffers in powers of 2. So outside of slices, we should only ever see a `1024` capacity or a `2048` capacity BB, nothing in between.
- I'd revert this change and go back to `>` because it's unrelated.
- let's use `.storageCapacity` even though in this case it's debatable I think

File: Sources/NIO/Codec.swift [hunk 2]
```diff
@@ -222,13 +222,13 @@ extension ByteToMessageDecoder {
     public func shouldReclaimBytes(buffer: ByteBuffer) -> Bool {
         // We want to reclaim in the following cases:
         //
-        // 1. If there is more than 2kB of memory to reclaim
+        // 1. If there is at least 2kB of memory to reclaim
         // 2. If the buffer is more than 50% reclaimable memory and is at least
         //    1kB in size.
-        if buffer.readerIndex > 2048 {
+        if buffer.readerIndex >= 2048 {
```
Comment (weissi) [inline Sources/NIO/Codec.swift:228]: what's the motivation for this change? Seems unrelated to the bugfix and obfuscates the important change?

File: Tests/NIOTests/CodecTest.swift [hunk 3]
```diff
@@ -210,19 +210,43 @@ public final class ByteToMessageDecoderTest: XCTestCase {
         _ = try channel.pipeline.addHandler(decoder).wait()
 
         // We're going to send in 513 bytes. This will cause a chunk to be passed on, and will leave
-        // a 512-byte empty region in a 513 byte buffer. This will not cause a shrink.
+        // a 512-byte empty region in a byte buffer with a capacity of 1024 bytes. Since 512 empty
```
Comment (weissi) [inline Tests/NIOTests/CodecTest.swift:213]: I think we should leave the old unit tests alone and add targeted new tests that specifically test for what was broken.

**Review by weissi — CHANGES_REQUESTED** [review]
Thanks very much! This is definitely a fix we want to have asap. Left some minor feedback.

File: Sources/NIO/Codec.swift (hunk 2 above)
Comment (fabianfett) [reply Sources/NIO/Codec.swift:228]: In `CodecTests` we currently explicitly test this behavior. The test succeeded up until now, since it used the buggy code path:
```swift
    func testMemoryIsReclaimedIfLotsIsAvailable() throws {
        let channel = EmbeddedChannel()
        defer {
            XCTAssertNoThrow(try channel.finish())
        }

        let decoder = ByteToMessageHandler(OnceDecoder()) // <--- the once decoder consumes exactly 2048 bytes.
        _ = try channel.pipeline.addHandler(decoder).wait()

        // We're going to send in 5119 bytes. This will be held.
        var buffer = channel.allocator.buffer(capacity: 5119)
        buffer.writeBytes(Array(repeating: 0x04, count: 5119))
        XCTAssertTrue(try channel.writeInbound(buffer).isEmpty)

        XCTAssertEqual(decoder.cumulationBuffer!.readableBytes, 5119)
        XCTAssertEqual(decoder.cumulationBuffer!.readerIndex, 0)

        // Now we're going to send in one more byte. This will cause a chunk to be passed on,
        // shrinking the held memory to 3072 bytes. However, memory will be reclaimed.
        XCTAssertTrue(try channel.writeInbound(buffer.getSlice(at: 0, length: 1)).isFull)
        XCTAssertEqual(decoder.cumulationBuffer!.readableBytes, 3072)
        XCTAssertEqual(decoder.cumulationBuffer!.readerIndex, 0)
    }
```

File: Tests/NIOTests/CodecTest.swift (hunk 3 above)
Comment (fabianfett) [reply Tests/NIOTests/CodecTest.swift:213]: With the fixed logic above this tests fails without modifications. For this reason, I adjusted it.

File: Sources/NIO/Codec.swift (hunk 1 above)
Comment (fabianfett) [reply Sources/NIO/Codec.swift]: Fixed.

File: Sources/NIO/Codec.swift (hunk 2 above)
Comment (weissi) [reply Sources/NIO/Codec.swift:228]: Thank you! That makes sense.

File: Tests/NIOTests/CodecTest.swift (hunk 3 above)
Comment (weissi) [reply Tests/NIOTests/CodecTest.swift:213]: Thank you! That makes sense.

File: Sources/NIO/Codec.swift (hunk 2 above)
Comment (weissi) [reply Sources/NIO/Codec.swift:228]: offline discussion with Fabian:

- the change in this logic is kinda arbitrary but it also won't hurt
- the documentation is just a code comment and _not_ in the public API docs so we never guaranteed what we'd do anyway so shifting the arbitrary limit by 1 byte shouldn't matter

**Review by weissi — APPROVED** [review]
Thank you! Really great fix!

File: Sources/NIO/Codec.swift (hunk 1 above)
Comment (weissi) [reply Sources/NIO/Codec.swift]: thank you

**Review by Lukasa — APPROVED** [review]
Nice, LGTM.
